bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,437 @@
+			log.Info("Machine upgrade failed for machine, requeuing...", "machine", m.Name)
+			if err := r.markUpgradeFailed(ctx, scope, m); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as failed: %w", err)
+			}
+
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		if r.isMachineUpgrading(m) {
+			log.Info("Machine is upgrading, requeuing...", "machine", m.Name)
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+
+		// Machine is not upgraded, mark it for upgrade
+		if err := r.markMachineToUpgrade(ctx, scope, m); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark machine to upgrade: %w", err)
+		}
+
+		log.Info("Machine marked for upgrade", "machine", m.Name)
+
+		if err := r.markUpgradeInProgress(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as in-progress: %w", err)
+		}
+	}
+
+	if upgradedMachines == len(ownedMachines) {
+		if err := r.markUpgradeDone(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as done: %w", err)
+		}
+
+		log.Info("All machines are upgraded")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+}
+
+// markUpgradeInProgress marks the MachineDeployment as in-place upgrade in-progress.
+func (r *MachineDeploymentReconciler) markUpgradeInProgress(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	ant := scope.MachineDeployment.Annotations
+	if ant == nil {
+		ant = make(map[string]string)
+	}
+
+	// clean up
+	delete(ant, bootstrapv1.InPlaceUpgradeReleaseAnnotation)
+
+	ant[bootstrapv1.InPlaceUpgradeToAnnotation] = scope.UpgradeTo
+	ant[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeInProgressStatus
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeInProgressEvent,
+		"MachineDeployment %s marked as in-place upgrade in-progress",
+		scope.MachineDeployment.Name,
+	)
+	return nil
+}
+
+// markUpgradeDone marks the MachineDeployment as in-place upgrade done.
+func (r *MachineDeploymentReconciler) markUpgradeDone(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	ant := scope.MachineDeployment.Annotations
+	if ant == nil {
+		ant = make(map[string]string)
+	}
+
+	// clean up
+	delete(ant, bootstrapv1.InPlaceUpgradeToAnnotation)
+
+	ant[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeDoneStatus
+	ant[bootstrapv1.InPlaceUpgradeReleaseAnnotation] = scope.UpgradeTo
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeDoneEvent,
+		"MachineDeployment %s marked as in-place upgrade done",
```

Comment:
This is the user-facing string, right? 
Let's have something more straight forward then:
`MachineDeployment in-place upgrade done`. 
No need to tell that we "marked" a resource, the user cares about the outcome. (same above)